addons/udes_advanced_packaging/models/stock_quant_package.py

```diff
@@ -101,3 +101,43 @@ def write(self, vals):
         """
         self.validate_parent_name(vals)
         return super().write(vals)
+
+    def has_same_content(self, other):
+        """Compare the content of current package with the content of another package."""
+        self.ensure_one()
+        return frozenset(self._get_all_products_quantities().items()) == frozenset(
+            other._get_all_products_quantities().items()
+        )
+
+    def _get_all_products_quantities(self):
+        """This function computes the different product quantities for the given package"""
+        # TDE CLEANME: probably to move somewhere else, like in pack op
+        res = {}
+        for quant in self._get_contained_quants():
+            if quant.product_id not in res:
+                res[quant.product_id] = 0
+            res[quant.product_id] += quant.quantity
+        return res
+
+    def assert_reserved_full_package(self, move_lines):
+        """Check that a package is fully reserved at move_lines."""
+        MoveLine = self.env["stock.move.line"]
+
+        self.ensure_one()
+
+        pack_products = frozenset(self._get_all_products_quantities().items())
+        mls_products = frozenset(move_lines._get_all_products_quantities().items())
+        if pack_products != mls_products:
+            # move_lines do not match the quants
+            picking = move_lines.mapped("picking_id")
+            picking.ensure_one()
+            pack_mls = MoveLine.search(
+                [("package_id", "child_of", self.id), ("state", "not in", ["done", "cancel"])]
+            )
+            other_pickings = pack_mls.mapped("picking_id") - picking
+            if other_pickings:
+                raise ValidationError(
+                    _("The package is reserved in other pickings: %s")
+                    % ",".join(other_pickings.mapped("name"))
+                )
+            raise ValidationError(_("Cannot mark as done a partially reserved package."))
```

Comment:
nit: Would it be useful to include the package name here in the error?

Reply:
I'm not convinced we should be adding this method as part of this story anyway.

On v11 it's used by the monster `update_picking` method, which we won't have in v14. But here we are only using it in a unit test.

Reply:
I think we can keep the function here, we might use it later on in other code. We won't have the monster function but we will have other functions that might need it.